src/Disks/ObjectStorages/MetadataStorageFromDiskTransactionOperations.cpp

```diff
@@ -287,6 +322,74 @@ void WriteFileOperation::undo()
     }
 }
 
+WriteFileAndCreateHardLinkOperation::WriteFileAndCreateHardLinkOperation(
+    const std::string & path_,
+    const std::string & index_path_,
+    IDisk & disk_,
+    const std::string & data_,
+    const MetadataStorageFromDisk & metadata_storage_
+    )
+    : path(path_)
+    , index_path(index_path_)
+    , disk(disk_)
+    , data(data_)
+    , metadata_storage(metadata_storage_)
+{
+}
+
+void WriteFileAndCreateHardLinkOperation::execute(std::unique_lock<std::shared_mutex> & lock)
+{
+    if (disk.exists(path))
+    {
+        existed = true;
+        auto buf = disk.readFile(path);
+        readStringUntilEOF(prev_data, *buf);
+    }
+
+    if (disk.exists(index_path))
+    {
+        // Another metadata for this object already exists
+        // So we create hardlink instead of new metadata file
+        auto metadata = metadata_storage.readMetadataUnlocked(index_path, lock);
+        metadata->incrementRefCount();
+        write_operation = std::make_unique<WriteFileOperation>(index_path, disk, metadata->serializeToString());
+        write_operation->execute(lock);
+        disk.createHardLink(index_path, path);
```

Comment:
Seems like the refcount written in metadata file and the actual number of hardlinks on filesystem may diverge if power is turned off

Reply:
Ok, but do we really need to write refcount into metadata file if we can always ask filesystem about actual number of hardlinks? Currently we have 3 different refcounts: `DiskObjectStorageMetadata::ref_count` in memory, refcount in metadata file and `st_nlink` on FS. And we have to keep them all consistent. Maybe we should remove refcount from metadata files, because it's redundant